### DISCUSSION TITLE
Don't Release Fluent2 With Resources in the Current State

### DISCUSSION THREAD

**Discussion by robert-abeo** [dotnet/wpf#9988] [General]
I've filed a number of issues related to the resources used in Fluent theme control templates. The resource structure and keys DO NOT follow upstream Fluent (in WinUI) for the most part. This is a problem because:

 1. Existing apps expect resources keys to be defined that aren't there. That prevents moving to the "official" WPF fluent theme here
 2. It appears the resource keys diverge here from upstream simply because the original author didn't know what they were doing. They didn't make things better, they broke a good design.
 3. Re-templating is required for a lot of cases where a lightweight styling solution is possible otherwise. Not enough resource keys are defined and used.

Every single Fluent control template I've looked at in depth has had this issue so far. It's causing real problems trying to adapt an existing app to use this theme.

The WPF team should NOT have done what they did here (applying minimal effort and importing these issues from the WPFUI project) and I think a course correction is needed. You must rebase the fluent theme on the existing resource keys in upstream WinUI. Use the same keys, the same 3-layer design (control-resources, resource brushes, resource colors... etc.). Everything. This will require a control-by-control review.

 1. Review each control by itself
 2. COPY the resources (and key names) from upstream WinUI for each control into this project. Including unused resource keys is fine as some apps may still use them.
 3. Adjust the copied resources to work in WPF. This can require some changes.
 4. Review the default style and control template in each control making sure it points to the correct resources. This will likely require handling a few more cases and adjusting triggers, etc.

The WPF team really needs to spend the time to fix this. The current state is pretty bad for a Microsoft product (comparing with WinUI another Microsoft product).

**Comment by pchaurasia14** [comment]
Hey @robert-abeo, thank you for sharing your analysis. We were anticipating few iterations of the Fluent theming before we get it exactly perfect. Rest assured, the problems you've mentioned above is at the top of our priority list and we'll get to fixing it as soon as possible.

**Reply by robert-abeo** [reply]
@pchaurasia14 Thanks, I appreciate the feedback.

I also keep in mind that in WinUI resource keys are considered part of the public API contract (Fluent was designed a bit differently compared to traditional WPF themes like Aero2). Changing resource keys can easily cause breaking changes in apps -- breaking changes that cause crashes at runtime.

As the bits are likely already finalized for .NET 9 just note an obsolete keys for compatibility section will be needed. It would have much easier and cleaner to change this before the initial release.

**Reply by miloush** [reply]
The API is marked as experimental, meaning it's not stable and can change.

**Reply by dipeshmsft** [reply]
@robert-abeo, regarding the obsolete keys for compatibility, we share this same thought, going forward while we fix the styles for each control, we aren't planning to remove any keys, so that we minimize the crashes at runtime. We will be starting to work on each control ASAP.

**Reply by miloush** [reply]
The keys are only accessible because the resources are merged in as a resource dictionary. As already discussed elsewhere, that is not appropriate for a theme.

**Reply by robert-abeo** [reply]
@miloush Respectfully, you don't know how Fluent theme was designed to be used. The XAML space has evolved since WPF -- we need those resource keys just like in UWP/WinUI/Avalonia.

**Reply by miloush** [reply]
Doesn't change that merging it in like normal styles is fundamentally broken. "Evolved" XAML also needs "evolved" infrastructure.

**Reply by robert-abeo** [reply]
Yes, but several 3rd party libraries have made it work with resource dictionaries, which is what we have.
